keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,758 @@
+
+This not only impacts kubectl, but all clients of kubernetes. We can do better.
+
+### Goals
+
+- Fix the discovery storm issue currently present in kubectl
+- Aggregate the discovery documents for all Kubernetes types
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Since the current discovery separated by group-version is already GA, removal of the endpoint will be difficult. This KEP will solely focus on introducing the new aggregated endpoint and will not cover deprecation.
+
+## Proposal
+
+We are proposing a new endpoint `/discovery` as an aggregated endpoint for all discovery documents. Discovery documents can currently be found under `apis/<group>/<version>` and `/api/v1` for the legacy group version. This discovery endpoint will support publishing an ETag so clients who already have the latest version of the aggregated discovery can avoid redownloading the document.
+
+We will add a new controller responsible for aggregating the discovery documents when a resource on the cluster changes. There will be no conflicts when aggregating since each discovery document is self-contained.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We will expose a endpoint `/discovery` that will support JSON, and
+protobuf. The endpoint will serve the aggregated discovery document
+for all types that a Kubernetes cluster supports.
+
+### API
+
+The contents of this endpoint will be an `APIGroupList`, which is the
```

Comment:
are you open to removing fields not related to discovery or that we have found limited use for?  `serverAddressByClientCIDRs` comes to mind as "wish we hadn't tried that"

Reply:
I'm assuming we can't deprecate the GA field and we would have to create a new struct without that field? I don't really have a strong preference, can add it to the KEP if we'd like to create new discovery structs.

Reply:
I think we should make our API serialization the type we want. I can understand if internally you have a RESTMapper convert to something else to make code easier to migrate.

Reply:
I agree with David, we can make a type that's a subset of the existing type and only includes necessary fields. We're not removing the existing pages, people can go there if they really need to.

Reply:
Did this make it into the merged KEP?  I don’t see the language in the merged version I expected (a new Kind, with fewer fields, and discussion of naming).

We have a mechanism to request a specific variant of a type (content negotiation), so I would prefer we use a kind that requires minimal changes and clarify what happens if you need those fields, not bifurcate the types.

Reply:
We're creating a new API in a new endpoint (`discovery/v1`) with a different API format compared to `/apis` and `apis/<group>/<version>`. The types are not round-trippable between the two APIs (eg: Subresource return types are changed, etc) and serve different endpoints which is why we ended up deciding to create new types.

Feel free to comment directly in the PR:  https://github.com/kubernetes/kubernetes/pull/111978